src/devices/cpu/z80/t6a84.cpp

```diff
@@ -0,0 +1,227 @@
+	irqack_cb().set([this](int state) {
+		LOGMASKED(LOG_PAGE_W, "IRQ ACK %02x => %02x\n", m_code_page, m_vector_page);
+		m_prev_code_page = m_code_page;
+		m_code_page = m_vector_page;
+	});
+	reti_cb().set([this](int state) {
+		LOGMASKED(LOG_PAGE_W, "IRQ RET %02x => %02x\n", m_code_page, m_prev_code_page);
+		m_code_page = m_prev_code_page;
+	});
+}
+
+void t6a84_device::device_start()
+{
+	z80_device::device_start();
+
+	space(AS_PROGRAM).invalidate_caches(read_or_write::READWRITE);
+	space(AS_DATA).invalidate_caches(read_or_write::READWRITE);
+
+	space(AS_PROGRAM).cache(m_args);
+	space(AS_DATA).specific(m_data);
+	space(AS_STACK).specific(m_stack);
+
+	save_item(NAME(m_code_page));
+	save_item(NAME(m_delay_code_page));
+	save_item(NAME(m_delay_code_pc));
+	save_item(NAME(m_is_delay_code_page_set));
+	save_item(NAME(m_prev_code_page));
+	save_item(NAME(m_data_page));
+	save_item(NAME(m_stack_page));
+	save_item(NAME(m_vector_page));
+}
+
+void t6a84_device::device_reset()
+{
+	m_code_page = 0;
+	m_delay_code_page = 0;
+	m_delay_code_pc = 0;
+	m_is_delay_code_page_set = false;
+	m_prev_code_page = 0;
+	m_data_page = 8;
+	m_stack_page = 8;
+	m_vector_page = 0;
+
+	z80_device::device_reset();
+}
+
+device_memory_interface::space_config_vector t6a84_device::memory_space_config() const
+{
+	auto r = z80_device::memory_space_config();
+	r.emplace_back(AS_DATA, &m_data_space_config);
+	r.emplace_back(AS_STACK, &m_stack_space_config);
+	for (auto it = r.begin(); it != r.end(); ++it) {
+		if ((*it).first == AS_IO) {
+			(*it).second = &m_io_space_config;
+		} else if ((*it).first == AS_PROGRAM) {
+			(*it).second = &m_program_space_config;
+		}
+	}
+
+	return r;
+}
+
+bool t6a84_device::memory_translate(int spacenum, int intention, offs_t &address, address_space *&target_space)
+{
+	if (spacenum == AS_PROGRAM) {
+		address = code_address(address);
+	} else if (spacenum == AS_DATA) {
+		address = data_address(address);
+	} else if (spacenum == AS_STACK) {
+		address = stack_address(address);
+	}
+
+	target_space = &space(spacenum);
+
+	return true;
+}
+
+uint32_t t6a84_device::code_address(uint16_t address)
+{
+	/*
+	    FIXME: Replace address range by actual address of next jump instruction.
+
+	    When setting a code page, it only becomes effective after jumping to a far address in that page.
+	    Any instructions fetched and executed before that jump still use the previous code page.
+	    This can be seen in Ferie Kitty, when test program at page 7 gets mapped, as we are still
+	    executing on page 0, but we expect to start executing that program when jumping to RST0:
+
+	    ROM_00::1ea9 3e 07      LD   A,0x7
+	    ROM_00::1eab d3 fe      OUT  (DAT_io_00fe),A
+	    ROM_00::1ead c3 00 00   JP   RST0
+	*/
+	if (!machine().side_effects_disabled() && m_is_delay_code_page_set && (address < m_delay_code_pc || address > (m_delay_code_pc + 0x20))) {
+		m_code_page = m_delay_code_page;
+		m_is_delay_code_page_set = false;
+	}
```

Comment:
Doing this on address translation seems wrong – it should happen on the jump instruction.  Does the Z80 base class allow customising handling for instructions?

Reply:
I didn't see any, but your suggestion makes sense. Maybe we can add a callback for branching instructions?